### PR TITLE
Information Retrieval (5/5)

### PR DESCRIPTION
**Pull request by lucadiliello**
# Before submitting

- [X] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure to update the docs?   
- [X] Did you write any new necessary tests?  

## What does this PR do?
Implements Normalized Discounted Cumulative Gain for IR. Original discussion [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/4991)

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]
lgtm

**Comment by Borda** [comment]
@lucadiliello mind adding an entry to changelog... 🐰

File: torchmetrics/functional/retrieval/ndcg.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        >>> from torchmetrics.functional import retrieval_normalized_dcg
+        >>> preds = torch.tensor([.1, .2, .3, 4, 70])
+        >>> target = torch.tensor([10, 0, 0, 1, 5])
+        >>> retrieval_normalized_dcg(preds, target)
+        tensor(0.6957)
+    """
+    preds, target = _check_retrieval_functional_inputs(preds, target, allow_non_binary_target=True)
+
+    if k is None:
+        k = preds.shape[-1]
```
Comment (SkafteNicki) [inline torchmetrics/functional/retrieval/ndcg.py:47]:
```suggestion
    k = preds.shape[-1] if k is None else k
```

File: torchmetrics/functional/retrieval/ndcg.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    if not (isinstance(k, int) and k > 0):
+        raise ValueError("`k` has to be a positive integer or None")
+
+    if not target.sum():
+        return tensor(0.0, device=preds.device)
+
+    sorted_target = target[torch.argsort(preds, dim=-1, descending=True)][:k]
+    ideal_target = torch.sort(target, descending=True)[0][:k]
+
+    def dcg(target):
```
Comment (SkafteNicki) [inline torchmetrics/functional/retrieval/ndcg.py:58]: please no local functions, move outside an rename `_dcg`

Comment (lucadiliello) [reply torchmetrics/functional/retrieval/ndcg.py:57]: Ok!

File: torchmetrics/functional/retrieval/ndcg.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import torch
+from torch import Tensor, tensor
+
+from torchmetrics.utilities.checks import _check_retrieval_functional_inputs
+
+
+def _dcg(target):
+        return (target / torch.log2(torch.arange(target.shape[-1]) + 2.0)).sum()
```
Comment (SkafteNicki) [inline torchmetrics/functional/retrieval/ndcg.py:21]:
```suggestion
    return (target / torch.log2(torch.arange(target.shape[-1]) + 2.0)).sum()
```

**Review by SkafteNicki — APPROVED** [review]

**Comment by Borda** [comment]
@lucadiliello mind resolve conflicts and fix the GPU failer :]

File: torchmetrics/functional/retrieval/ndcg.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import torch
+from torch import Tensor, tensor
+
+from torchmetrics.utilities.checks import _check_retrieval_functional_inputs
+
+
+def _dcg(target):
+    return (target / torch.log2(torch.arange(target.shape[-1]) + 2.0)).sum()
```
Comment (SkafteNicki) [inline torchmetrics/functional/retrieval/ndcg.py:21]:
```suggestion
    denom = torch.log2(torch.arange(target.shape[-1], device=target.device) + 2.0)
    return (target / denom).sum()
```
